### PR TITLE
Bump version to 16.0.3 (C++17 support)

### PR DESCRIPTION
**Pull request by andreittr**
Updated libcxxabi to the latest upstream version; based off earlier work from @mschlumpp .
Part of a patch set w/ further PRs under [libcxx](https://github.com/unikraft/lib-libcxx/pull/27), [lib-compiler-rt](https://github.com/unikraft/lib-compiler-rt/pull/11), [libunwind](https://github.com/unikraft/lib-libunwind/pull/6) & [lib-musl](https://github.com/unikraft/lib-musl/pull/45).

### CODE REVIEW

**Review by StefanJum — CHANGES_REQUESTED** [review]
Hi @andreittr, same as in the [`lib-compiler-rt` version bump pr](https://github.com/unikraft/lib-compiler-rt/pull/11):

Please change the commits to match the [unikraft commit format](https://unikraft.org/docs/contributing/submitting-changes/#commits). (i.e. sign-off them and add the co-authored-by and sign-off of the co-authors, add the component/selector to the commit message).

Besides that, the build of [`helloworld-cpp`](https://github.com/unikraft/app-helloworld-cpp) fails with:
```console
In file included from /media/stefan/projects/unikraft/scripts/workdir/apps/app-helloworld-cpp/build/libcxxabi/origin/libcxxabi-16.0.3.src/src/cxa_guard.cpp:15:
/media/stefan/projects/unikraft/scripts/workdir/apps/app-helloworld-cpp/build/libcxxabi/origin/libcxxabi-16.0.3.src/src/cxa_guard_impl.h:627:1: error: ‘_LIBCPP_CONSTINIT’ does not name a type
  627 | _LIBCPP_CONSTINIT T GlobalStatic<T>::instance = {};
      | ^~~~~~~~~~~~~~~~~
cc1plus: warning: command-line option ‘-fhosted’ is valid for C/ObjC but not for C++
make[3]: *** [/media/stefan/projects/unikraft/scripts/workdir/unikraft/support/build/Makefile.build:27: /media/stefan/projects/unikraft/scripts/workdir/apps/app-helloworld-cpp/build/libcxxabi/cxa_guard.o] Erro1 
make[3]: *** Waiting for unfinished jobs....
cc1plus: warning: command-line option ‘-fhosted’ is valid for C/ObjC but not for C++
/media/stefan/projects/unikraft/scripts/workdir/apps/app-helloworld-cpp/build/libcxxabi/origin/libcxxabi-16.0.3.src/src/fallback_malloc.cpp:39:8: error: ‘_LIBCPP_CONSTINIT’ does not name a type
   39 | static _LIBCPP_CONSTINIT void* heap_mutex = 0;
      |        ^~~~~~~~~~~~~~~~~
/media/stefan/projects/unikraft/scripts/workdir/apps/app-helloworld-cpp/build/libcxxabi/origin/libcxxabi-16.0.3.src/src/fallback_malloc.cpp: In function ‘void* {anonymous}::fallback_malloc(size_t)’:
/media/stefan/projects/unikraft/scripts/workdir/apps/app-helloworld-cpp/build/libcxxabi/origin/libcxxabi-16.0.3.src/src/fallback_malloc.cpp:134:16: error: ‘heap_mutex’ was not declared in this scope
  134 |   mutexor mtx(&heap_mutex);
      |                ^~~~~~~~~~
/media/stefan/projects/unikraft/scripts/workdir/apps/app-helloworld-cpp/build/libcxxabi/origin/libcxxabi-16.0.3.src/src/fallback_malloc.cpp: In function ‘void {anonymous}::fallback_free(void*)’:
/media/stefan/projects/unikraft/scripts/workdir/apps/app-helloworld-cpp/build/libcxxabi/origin/libcxxabi-16.0.3.src/src/fallback_malloc.cpp:193:16: error: ‘heap_mutex’ was not declared in this scope
  193 |   mutexor mtx(&heap_mutex);
      |                ^~~~~~~~~~
/media/stefan/projects/unikraft/scripts/workdir/apps/app-helloworld-cpp/build/libcxxabi/origin/libcxxabi-16.0.3.src/src/fallback_malloc.cpp: In function ‘void* __cxxabiv1::__aligned_malloc_with_fallback(size_t: 
/media/stefan/projects/unikraft/scripts/workdir/apps/app-helloworld-cpp/build/libcxxabi/origin/libcxxabi-16.0.3.src/src/fallback_malloc.cpp:267:25: error: ‘__libcpp_aligned_alloc’ is not a member of ‘std’
  267 |   if (void* dest = std::__libcpp_aligned_alloc(__alignof(__aligned_type), size))
      |                         ^~~~~~~~~~~~~~~~~~~~~~
/media/stefan/projects/unikraft/scripts/workdir/apps/app-helloworld-cpp/build/libcxxabi/origin/libcxxabi-16.0.3.src/src/fallback_malloc.cpp: In function ‘void __cxxabiv1::__aligned_free_with_fallback(void*)’:
/media/stefan/projects/unikraft/scripts/workdir/apps/app-helloworld-cpp/build/libcxxabi/origin/libcxxabi-16.0.3.src/src/fallback_malloc.cpp:291:10: error: ‘__libcpp_aligned_free’ is not a member of ‘std’
  291 |     std::__libcpp_aligned_free(ptr);
      |          ^~~~~~~~~~~~~~~~~~~~~
make[3]: *** [/media/stefan/projects/unikraft/scripts/workdir/unikraft/support/build/Makefile.build:27: /media/stefan/projects/unikraft/scripts/workdir/apps/app-helloworld-cpp/build/libcxxabi/fallback_malloc.oE1
make[2]: *** [Makefile:1032: sub-make] Error 2
make[1]: *** [Makefile:32: _all] Error 2
make: *** [Makefile:7: all] Error 2
```

**Comment by andreittr** [comment]
Hi @StefanJum !

> Besides that, the build of [helloworld-cpp](https://github.com/unikraft/app-helloworld-cpp) fails with:

Are you building against the [similarly updated `libcxx`](https://github.com/unikraft/lib-libcxx/pull/27)? The version in staging is much older (v7.0.0) and likely does not provide everything this newer libcxxabi expects.

**Comment by StefanJum** [comment]
> Are you building against the [similarly updated `libcxx`](https://github.com/unikraft/lib-libcxx/pull/27)? The version in staging is much older (v7.0.0) and likely does not provide everything this newer libcxxabi expects.

Ah, I was not aware of that PR, thanks @andreittr.
The build environment I'm testing now is everything on `staging` + the following PRs:
* [compiler-rt/11](https://github.com/unikraft/lib-compiler-rt/pull/11)
* [libcxx/27](https://github.com/unikraft/lib-libcxx/pull/27)
* [libcxxabi/3](https://github.com/unikraft/lib-libcxxabi/pull/3)
* [musl/44](https://github.com/unikraft/lib-musl/pull/44)
Is that all or should anything else be included?

Still with all this, the build fails but with different errors, all of them seem related to `libcxx`.
```console
.../apps/app-helloworld-cpp/build/libcxx/origin/libcxx-16.0.3.src/include/cctype:43:5: error: #error <cctype> tried including <ctype.h> but didn't find libc++'s <ctype.h> header. This usually means that your header search paths are not configured properly. The header search paths should contain the C++ Standard Library headers before any C Standard Library.
   43 | #   error <cctype> tried including <ctype.h> but didn't find libc++'s <ctype.h> header. \
      |     ^~~~~
In file included from .../apps/app-helloworld-cpp/build/libcxx/origin/libcxx-16.0.3.src/include/__locale:16,
                 from .../apps/app-helloworld-cpp/build/libcxx/origin/libcxx-16.0.3.src/include/ios:221,
                 from .../apps/app-helloworld-cpp/build/libcxx/origin/libcxx-16.0.3.src/include/iostream:43,
                 from .../apps/app-helloworld-cpp/helloworld.cpp:1:
.../apps/app-helloworld-cpp/build/libcxx/origin/libcxx-16.0.3.src/include/cstdint:149:5: error: #error <cstdint> tried including <stdint.h> but didn't find libc++'s <stdint.h> header. This usually means that your header search paths are not configured properly. The header search paths should contain the C++ Standard Library headers before any C Standard Library, and you are probably using compiler flags that make that not be the case.
  149 | #   error <cstdint> tried including <stdint.h> but didn't find libc++'s <stdint.h> header. \
      |     ^~~~~
.../apps/app-helloworld-cpp/build/libcxx/origin/libcxx-16.0.3.src/include/__memory/unique_ptr.h:581:23: error: expected primary-expression before ‘const’
  581 | operator==(nullptr_t, const unique_ptr<_T1, _D1>& __x) _NOEXCEPT
      |                       ^~~~~
.../apps/app-helloworld-cpp/build/libcxx/origin/libcxx-16.0.3.src/include/__memory/unique_ptr.h:581:48: error: expected primary-expression before ‘>’ token
  581 | operator==(nullptr_t, const unique_ptr<_T1, _D1>& __x) _NOEXCEPT
      |                                                ^
.../apps/app-helloworld-cpp/build/libcxx/origin/libcxx-16.0.3.src/include/cmath:343:9: error: ‘isgreaterequal’ has not been declared in ‘::’
  343 | using ::isgreaterequal _LIBCPP_USING_IF_EXISTS;
      |         ^~~~~~~~~~~~~~
```
There are a lot of them, all along the same lines as the ones above.

**Comment by andreittr** [comment]
> Still with all this, the build fails but with different errors, all of them seem related to libcxx.

Did you include musl in the build? I'm getting similar errors if I remove it from menuconfig.

Since libcxx depends on a somewhat complete libc, I can add LIBMUSL as a Kconfig select for libcxx. However, if we're thinking about supporting other libcs in the future this will need reworking. Thoughts?

**Comment by StefanJum** [comment]
> Since libcxx depends on a somewhat complete libc, I can add LIBMUSL as a Kconfig select for libcxx. However, if we're thinking about supporting other libcs in the future this will need reworking. Thoughts?

`musl` is selected in the menuconfig. What is your libraries order in the `Makefile`?

I would add `musl` it in the `Config.uk` only if it is a strict dependency (it's worth checking if it builds with the updated [`newlib` version](https://github.com/unikraft/lib-newlib/pull/28)).
We can probably do something like [`depends on HAVE_LIBC`](https://github.com/unikraft/unikraft/blob/staging/lib/Config.uk#L12-L14), I think it's cleaner, since `HAVE_LIBC` will be selected by any future ported libc.

**Comment by andreittr** [comment]
> What is your libraries order in the Makefile?

```
LIBS-y := $(UK_LIBS)/lib-libcxxabi:$(UK_LIBS)/lib-libcxx:$(UK_LIBS)/lib-compiler-rt:$(UK_LIBS)/lib-libunwind:$(UK_LIBS)/lib-musl
```
I was not even aware it mattered.

> We can probably do something like [depends on HAVE_LIBC](https://github.com/unikraft/unikraft/blob/staging/lib/Config.uk#L12-L14), I think it's cleaner, since HAVE_LIBC will be selected by any future ported libc.

I think that would be best; Kconfig will generate a warning if the dependency isn't fulfilled, so it won't fail silently. Will add this change to the libcxx PR.

**Comment by StefanJum** [comment]
> ```
> LIBS-y := $(UK_LIBS)/lib-libcxxabi:$(UK_LIBS)/lib-libcxx:$(UK_LIBS)/lib-compiler-rt:$(UK_LIBS)/lib-libunwind:$(UK_LIBS)/lib-musl
> ```
> 
> I was not even aware it mattered.

Yeah, unfortunately it affects the headers include ordering.
I changed it to what you provided and the errors changed again :smile: 
```
In file included from workdir/apps/app-helloworld-cpp/build/libcxx/origin/libcxx-16.0.3.src/include/__algorithm/copy.h:16,
                 from workdir/apps/app-helloworld-cpp/build/libcxx/origin/libcxx-16.0.3.src/include/__memory/uninitialized_algorithms.h:13,
                 from workdir/apps/app-helloworld-cpp/build/libcxx/origin/libcxx-16.0.3.src/include/__memory/shared_ptr.h:30,
                 from workdir/apps/app-helloworld-cpp/build/libcxx/origin/libcxx-16.0.3.src/include/mutex:191,
                 from workdir/apps/app-helloworld-cpp/build/libcxx/origin/libcxx-16.0.3.src/include/__locale:18,
                 from workdir/apps/app-helloworld-cpp/build/libcxx/origin/libcxx-16.0.3.src/include/ios:221,
                 from workdir/apps/app-helloworld-cpp/build/libcxx/origin/libcxx-16.0.3.src/include/iostream:43,
                 from workdir/apps/app-helloworld-cpp/helloworld.cpp:1:
apps/app-helloworld-cpp/build/libcxx/origin/libcxx-16.0.3.src/include/__iterator/segmented_iterator.h:72:46: error: template argument ‘(sizeof (_Tp) * 0)’ involve)
   72 | struct __has_specialization<_Tp, sizeof(_Tp) * 0> : true_type {};
      |                                  ~~~~~~~~~~~~^~~
In file included from workdir/apps/app-helloworld-cpp/build/libcxx/origin/libcxx-16.0.3.src/include/__mutex_base:20,
                 from workdir/apps/app-helloworld-cpp/build/libcxx/origin/libcxx-16.0.3.src/include/mutex:192,
                 from workdir/apps/app-helloworld-cpp/build/libcxx/origin/libcxx-16.0.3.src/include/__locale:18,
                 from workdir/apps/app-helloworld-cpp/build/libcxx/origin/libcxx-16.0.3.src/include/ios:221,
                 from workdir/apps/app-helloworld-cpp/build/libcxx/origin/libcxx-16.0.3.src/include/iostream:43,
                 from workdir/apps/app-helloworld-cpp/helloworld.cpp:1:
workdir/apps/app-helloworld-cpp/build/libcxx/origin/libcxx-16.0.3.src/include/system_error: In instantiation of ‘std::__1::error_code::error_code(_Ep, typename std::__1:::
workdir/apps/app-helloworld-cpp/build/libcxx/origin/libcxx-16.0.3.src/include/ios:452:34:   required from here
workdir/apps/app-helloworld-cpp/build/libcxx/origin/libcxx-16.0.3.src/include/system_error:355:40: error: use of deleted function ‘void std::__1::__adl_only::make_error_c’
  355 |                 *this = make_error_code(__e);
      |                         ~~~~~~~~~~~~~~~^~~~~
workdir/apps/app-helloworld-cpp/build/libcxx/origin/libcxx-16.0.3.src/include/system_error:263:10: note: declared here
  263 |     void make_error_code() = delete;

```

**Comment by andreittr** [comment]
> I changed it to what you provided and the errors changed again

What compiler version are you using? I'm trying to get to the bottom of this but have a hard time replicating it :/

**Comment by StefanJum** [comment]
> > I changed it to what you provided and the errors changed again
> 
> What compiler version are you using? I'm trying to get to the bottom of this but have a hard time replicating it :/

`gcc version 11.1.0 (Ubuntu 11.1.0-1ubuntu1~20.04)`
I can put together a repo with the entire setup later today, I'll let you know when it's done.

**Comment by andreittr** [comment]
> gcc version 11.1.0 (Ubuntu 11.1.0-1ubuntu1~20.04)

Just built gcc 11 and can reproduce the exact same error on my machine => might be an issue with missing feature support in older versions; gcc 13 builds it successfully. Will try more versions and browse gcc changes to see where the cutoff point is.

**Comment by StefanJum** [comment]
> Just built gcc 11 and can reproduce the exact same error on my machine => might be an issue with missing feature support in older versions; gcc 13 builds it successfully. Will try more versions and browse gcc changes to see where the cutoff point is.

Got it, I'll also try gcc 13 and see if it works.

**Comment by andreittr** [comment]
Found a gcc bug related to the template argument error: https://gcc.gnu.org/bugzilla/show_bug.cgi?id=96555 ; seems to be fixed in gcc 12 (checked w/ the godbolt link in the bug comments).

**Comment by StefanJum** [comment]
> Found a gcc bug related to the template argument error: https://gcc.gnu.org/bugzilla/show_bug.cgi?id=96555 ; seems to be fixed in gcc 12 (checked w/ the godbolt link in the bug comments).

Hmm, I see. I would really not merge this if it breaks all builds with gcc < 12, since a lot of people do use older compiler versions.
Maybe we can [check the gcc version](https://github.com/unikraft/unikraft/blob/staging/support/build/Makefile.rules#L44-L46) at build-time and get around the issue somehow.

**Comment by andreittr** [comment]
> Maybe we can [check the gcc version](https://github.com/unikraft/unikraft/blob/staging/support/build/Makefile.rules#L44-L46) at build-time and get around the issue somehow.

I'm not sure it would be easy to patch libcxx to work around this; for now I'll have a look whether we can get decent C++17 support with an older libcxx that compiles with gcc 11 or earlier.
A more flexible solution would be to support multiple lib versions or multiple c++ stdlibs in the future.

**Comment by StefanJum** [comment]
> I'm not sure it would be easy to patch libcxx to work around this; for now I'll have a look whether we can get decent C++17 support with an older libcxx that compiles with gcc 11 or earlier. A more flexible solution would be to support multiple lib versions or multiple c++ stdlibs in the future.

I agree multiple library versions would be really nice, library versioning has been a topic on the discussions list for a long time (multiple c++ stdlibs even nicer :rocket: ). 
We should get to discuss that after the 0.13 release.

**Comment by andreittr** [comment]
Building libcxx v16 requires GCC >= 12 and that is a bit too restrictive a change for now.
I'm going to close this set of PRs and resubmit new ones that bump LLVM only to 14, along with the minimal necessary changes to get it working. libcxx 14 _does_ compile with GCC 11 (though not with earlier versions :( ).

**Pull request closed without merging**
